indexer/lib/advertisement-walker.js

```diff
@@ -252,23 +260,43 @@ export async function fetchAdvertisedPayload (providerAddress, advertisementCid,
     return { previousAdvertisementCid }
   }
 
-  const entriesChunk =
+  const meta = parseMetadata(advertisement.Metadata['/'].bytes)
+  const pieceCid = meta.deal?.PieceCID.toString()
+
+  try {
+    const entriesChunk =
     /** @type {{
      Entries: { '/' :  { bytes: string } }[]
     }} */(
-      await fetchCid(providerAddress, entriesCid, { fetchTimeout })
-    )
-  debug('entriesChunk %s %j', entriesCid, entriesChunk.Entries.slice(0, 5))
-  const entryHash = entriesChunk.Entries[0]['/'].bytes
-  const payloadCid = CID.create(1, 0x55 /* raw */, multihash.decode(Buffer.from(entryHash, 'base64'))).toString()
+        await fetchCid(providerAddress, entriesCid, { fetchTimeout })
+      )
+    debug('entriesChunk %s %j', entriesCid, entriesChunk.Entries.slice(0, 5))
+    const entryHash = entriesChunk.Entries[0]['/'].bytes
+    const payloadCid = CID.create(1, 0x55 /* raw */, multihash.decode(Buffer.from(entryHash, 'base64'))).toString()
 
-  const meta = parseMetadata(advertisement.Metadata['/'].bytes)
-  const pieceCid = meta.deal?.PieceCID.toString()
-
-  return {
-    previousAdvertisementCid,
-    pieceCid,
-    payloadCid
+    return {
+      previousAdvertisementCid,
+      pieceCid,
+      payloadCid
+    }
+  } catch (err) {
+    if (err && typeof err === 'object' && 'statusCode' in err && err.statusCode === 404) {
+      // The index provider cannot find the advertised entries. We cannot do much about that,
+      // it's unlikely that further request will succeed. Let's skip this advertisement.
+      debug(
+        'Cannot fetch ad %s entries %s: %s %s',
+        advertisementCid,
+        entriesCid,
+        err.statusCode,
+        /** @type {any} */(err).serverMessage ?? '<not found>'
+      )
+      return {
```

Comment:
Why are we returning an object with `{ error }` instead of throwing it?

Reply:
Yeah, this can be confusing, and I am thinking about redesigning my implementation.

What I wanted to accomplish in this PR:
- If the request for entries returns 404, we want to skip the advertisement and walk to the next advertisement in the chain. This is handled by returning `{ error }`
- If the request for entries fails with a different error, we want to retry it. This is handled by (re)throwing the error.

After thinking about this some more, I am leaning towards the following design:

- If the request for entries returns 404, we skip the advertisement. (No change from the current proposal.)
- If the request for entries fails differently, we retry it several times. If all attempts fail, then we skip the advertisement, too.

That way, we optimise for finishing the walk of all advertisements so that we can process new advertisements and avoid repeating the same failing request forever.